wgpu-hal/src/lib.rs

```diff
@@ -939,6 +940,25 @@ pub struct SamplerDescriptor<'a> {
     pub border_color: Option<wgt::SamplerBorderColor>,
 }
 
+impl Hash for SamplerDescriptor<'_> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.label.hash(state);
```

Comment:
Does that mean there's no possibility that we break the use case of the user creating multiple samplers with different labels expecting to get their unique labels back in the wgpu errors? If you're not 100% sure, I'd be happy to double check it, but it also means that dx12 isn't behaving as expected, no? Feels like it might change in the future.

Reply:
Tbh, I'm not sure if we have to give the labels to the backends in general, and we only do because it makes it easier to debug in case we missed something in wgpu-core and the backend itself is erroring.

For the dx12 backend, I've seen buffers and textures use `.SetName` the assign the label but I'm not sure if there is a way to do it for samplers.

@cwfitzgerald is it true that we only give labels to the underlying backends just in case our validation didn't catch an error? And theoretically we could only do it in debug builds?

Reply:
They're there for two reasons: validation and external tools. Things like renderdoc, PIX, and profilers all use them.

We've long only set names in debug mode, but I think that's a mistake, that's not a decision we get to make.

Reply:
Especially as debug labels are quite useful in profilers

Reply:
I feel like it's best to include the label in the hash then and add a comment to the hash function telling them not to add too many different labels

Reply:
@teoxoy I still think it could make sense to move the hasher into the dx12 backend, makes it more clear that it's only being used there.

Reply:
> We've long only set names in debug mode, but I think that's a mistake, that's not a decision we get to make.

I see, ok.

> I feel like it's best to include the label in the hash then and add a comment to the hash function telling the user not to add too many different labels

For sure, if we pass them to dx12, but as of right now I couldn't find where/if we do that.

Reply:
Ok when I pass over this again I'll try to get a definitive answer to that

Reply:
> I feel like it's best to include the label in the hash then and add a comment to the hash function telling the user not to add too many different labels

Here I actually disagree - with samplers being _such_ a scarce resource, we should try to de-duplicate as much as possible. Especially when the cause could be completely unintuitive.

In an ideal world, we could make the label assigned to the d3d resource the union of the names of all of the actual wgpu samplers using it. That means when you get a cache hit, you append the new label onto the existing sampler.